scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantBraces.scala

```diff
@@ -1,87 +1,143 @@
-        val open = d.body.tokens.head
-        val close = d.body.tokens.last
-        val bodyStatement = d.body match {
-          case t: Term.Block => t.stats.head
-          case _ => d.body
+        processInterpolation(t)
+    }
+
+    builder.result()
+  }
+
+  private def removeTrailingLF(bodyEnd: Position, close: Token)(
+      implicit builder: PatchBuilder,
+      ctx: RewriteCtx): Unit =
+    if (close.pos.start.line != bodyEnd.end.line) {
+      import ctx.tokenTraverser._
+      val next = nextToken(close)
+      if (next.is[LF])
+        builder += TokenPatch.Remove(next)
+    }
+
+  private def processBlock(
+      b: Term.Block)(implicit builder: PatchBuilder, ctx: RewriteCtx): Unit =
+    if (b.tokens.nonEmpty) {
+      val open = b.tokens.head
+      if (open.is[LeftBrace]) {
+        val close = b.tokens.last
+        if (removeBlock(b) && close.is[RightBrace]) {
+          val endPos = if (b.stats.isEmpty) b.pos else b.stats.last.pos
+          removeTrailingLF(endPos, close)
+          builder += TokenPatch.Remove(open)
+          builder += TokenPatch.Remove(close)
         }
-        val next = nextToken(close)
-        if (next.is[LF] &&
-          close.pos.start.line != bodyStatement.pos.end.line)
-          builder += TokenPatch.Remove(next)
+      }
+    }
+
+  private def removeBlock(b: Term.Block)(implicit ctx: RewriteCtx): Boolean = {
+    def exactlyOneStatement = b.stats.lengthCompare(1) == 0
+    b.parent.exists {
+
+      case _: Case => true
+
+      case _: Term.Apply => false // Leave this alone (for now...)
 
-        builder += TokenPatch.Remove(open)
-        builder += TokenPatch.Remove(close)
+      case d: Defn.Def =>
+        def disqualifiedByUnit =
+          !settings.includeUnitMethods && d.decltpe.exists(_.syntax == "Unit")
+        def innerOk =
+          b.stats.head match {
+            case _: Term.Function | _: Defn => false
+            case _ => true
+          }
+        exactlyOneStatement && blockSizeIsOk(b) && innerOk && !isProcedureSyntax(
+          d) && !disqualifiedByUnit
+
+      case _ =>
+        exactlyOneStatement && blockSizeIsOk(b) && !retainSingleStatBlock(b)
     }
-    builder.result()
   }
+
+  /** Some blocks look redundant but aren't */
+  private def retainSingleStatBlock(b: Term.Block): Boolean =
+    b.parent.exists {
+      case parentIf: Term.If =>
+        // if (a) { if (b) c } else d
+        //   ↑ cannot be replaced by ↓
+        // if (a) if (b) c else d
+        //   which would be equivalent to
+        // if (a) { if (b) c else d }
+        def insideIfThen = parentIf.thenp eq b
+        def parentIfHasAnElse = parentIf.elsep.tokens.nonEmpty
+        def blockIsIfWithoutElse = b.stats.head match {
+          case childIf: Term.If => childIf.elsep.tokens.isEmpty
+          case _ => false
+        }
+        insideIfThen && parentIfHasAnElse && blockIsIfWithoutElse
+
+      case _ =>
+        // No other (known) special cases
```

Comment:
When I think of it, I'm afraid special cases include whenever blocks are used to influence precedence

```diff
- {1 + 2} * 3
+ 1 + 2 * 3
- case x if { x match { case 1 => true } }
+ case x if x match { case 1 => true }
```

It's possible to detect these cases, but it requires a bit of machinery. 

* label all tree nodes into syntactic groups https://github.com/scalacenter/scala-syntax/blob/f7a6206aa95756855132d3de4e980ace6944a3c5/format/src/main/scala/org/scalafmt/internal/TreeSyntacticGroup.scala
* compare two syntactic groups https://github.com/scalacenter/scala-syntax/blob/f7a6206aa95756855132d3de4e980ace6944a3c5/format/src/main/scala/org/scalafmt/internal/TreeOps.scala#L58

Reply:
I took the liberty to integrate those syntactic groups to ensure redundant braces does not remove blocks that are required to influence precedence and associativity.